runtime/vm/options.c

```diff
@@ -0,0 +1,198 @@
+	"{default}",
+	"{command line}",
+	"{ergonomic}"
+};
+#define NUM_HOWSET_STRINGS	(sizeof(option_howset_string) / sizeof(option_howset_string[0]))
+
+typedef struct J9OptionDesc {
+	const char * name;
+	I_8 type;
+	I_8 category;
+	I_8 howset;
+	J9OptionValue value;
+} J9OptionDesc;
+
+
+/* options_enum gives each option an index starting from zero */
+typedef enum options_enum {
+	NoOption=-1,
+
+	#define OPTION(n,t,cats,how) IDX_ ## n,
+	#include "options.inc"
+	#undef OPTION
+} options_enum;
+
+/* options array lists all options, same order as options_enum */
+static J9OptionDesc options[]={
+	#define OPTION(n,t,cats,how) { n, t, cats, how, { 0 } },
+	#include "options.inc"
+	#undef OPTION
+};
+#define NUM_OPTIONS	(sizeof(options)/sizeof(options[0]))
+
+
+/*
+ * initializeFlagSettings initializes the value and howset fields for all options in
+ *	the options array. Constant default values would get howset=SET_DEFAULT, values
+ *	that are computed heuristically based on other options or the environment get
+ *	howset=SET_ERGONOMIC, and if the value was set via a command line option (including
+ *	aliases like -XX::MaxMemorySize= and -Xmx) get howset=SET_COMMANDLINE .
+ */
+void initializeFlagSettings(J9JavaVM* vm) {
+	PORT_ACCESS_FROM_JAVAVM(vm);
+	J9OptionDesc *option=NULL;
+
+	/* Initialize value for -XX:MaxHeapSize / -Xmx */
+	option = options + IDX_VMOPT_MAXHEAPSIZE;
+	option->value.v_sizet = vm->memoryManagerFunctions->j9gc_get_maximum_heap_size(vm);
+	if ((findArgInVMArgs( PORTLIB, vm->vmArgsArray, STARTSWITH_MATCH, VMOPT_XMX, NULL, 0) >= 0))
+		option->howset = SET_COMMANDLINE;
+	else if ((findArgInVMArgs( PORTLIB, vm->vmArgsArray, STARTSWITH_MATCH, VMOPT_XXMAXHEAPSIZE, NULL, 0) >= 0))
+		option->howset = SET_COMMANDLINE;
+
+	/* Initialize value for -XX:MaxDirectMemorySize */
+	option = options + IDX_VMOPT_MAXDIRECTMEMORYSIZE;
+	option->value.v_uint64 = vm->directByteBufferMemoryMax;
+	if ((findArgInVMArgs( PORTLIB, vm->vmArgsArray, STARTSWITH_MATCH, VMOPT_XXMAXDIRECTMEMORYSIZE, NULL, 0) >= 0))
+		option->howset = SET_COMMANDLINE;
+}
+
+/*
+ * Prints flags (options) to the console. The list of options printed is not yet complete.
+ *	See example output comment at the top of this file below the copyright header.
+ */
+void printFlags(J9JavaVM* vm) {
+	PORT_ACCESS_FROM_JAVAVM(vm);
+	UDATA o_idx=0;
+
+	j9tty_printf(PORTLIB, "[Global flags]\n");
+	for (o_idx=0; o_idx < NUM_OPTIONS; o_idx++) {
+		J9OptionDesc *option = options + o_idx;
+
+		assert(option->type < NUM_TYPE_STRINGS);
+		j9tty_printf(PORTLIB, "%9s %-40s = ", option_type_string[option->type], option->name);
+
+		switch (option->type) {
+			case TYPE_SIZET :
+				j9tty_printf(PORTLIB, "%-36zu", (size_t)(option->value.v_sizet));
+				break;
+
+			case TYPE_UINT64T :
+				j9tty_printf(PORTLIB, "%-36llu", (uint64_t)(option->value.v_uint64));
+				break;
+			default :
+				/* new option must have been added with an as yet unsupported type */
+				assert(0);
+		}
```

Comment:
Formatting nitpick - we typically don't indent the cases of a switch statement to avoid excessively long lines
```suggestion
		switch (option->type) {
		case TYPE_SIZET :
			j9tty_printf(PORTLIB, "%-36zu", (size_t)(option->value.v_sizet));
			break;

		case TYPE_UINT64T :
			j9tty_printf(PORTLIB, "%-36llu", (uint64_t)(option->value.v_uint64));
			break;
		default :
			/* new option must have been added with an as yet unsupported type */
			assert(0);
		}
```